framework/src/base/Assembly.C

```diff
@@ -3331,6 +3331,32 @@ Assembly::addCachedJacobian(SparseMatrix<Number> & /*jacobian*/)
   addCachedJacobian();
 }
 
+Real
+Assembly::elementVolume(const Elem * elem)
+{
+  FEType fe_type(elem->default_order(), LAGRANGE);
+  std::unique_ptr<FEBase> fe(FEBase::build(elem->dim(), fe_type));
+
+  const std::vector<Real> & JxW = fe->get_JxW();
+  // The default quadrature rule should integrate the mass matrix,
+  // thus it should be plenty to compute the volume
+  QGauss qrule(elem->dim(), fe_type.default_quadrature_order());
+  MooseArray<Point> q_points;
+  MooseArray<Real> coord;
+
+  fe->attach_quadrature_rule(&qrule);
+  fe->reinit(elem);
+
+  setCoordinateTransformation<ComputeStage::RESIDUAL>(
+      &qrule, q_points, coord, elem->subdomain_id());
+  coord.resize(qrule.n_points());
+  Real vol = 0;
+  for (unsigned int qp = 0; qp < qrule.n_points(); ++qp)
+    vol += JxW[qp] * coord[qp];
+
+  return vol;
+}
```

Comment:
This function is actually quite similar as `Elem::volume()` in libMesh. I think we possibly can enhance that function to accept coordinate type as an argument and add the following piece of codes:
```
  switch (_coord_type)
  {
    case Moose::COORD_XYZ:
      for (unsigned int qp = 0; qp < n_points; qp++)
        coord[qp] = 1.;
      break;

    case Moose::COORD_RZ:
      for (unsigned int qp = 0; qp < n_points; qp++)
        coord[qp] = 2 * M_PI * q_points[qp](rz_radial_coord);
      break;

    case Moose::COORD_RSPHERICAL:
      for (unsigned int qp = 0; qp < n_points; qp++)
        coord[qp] = 4 * M_PI * q_points[qp](0) * q_points[qp](0);
      break;

    default:
      mooseError("Unknown coordinate system");
      break;
  }
```

Reply:
But that doesn't help if the instance of MOOSE code doesn't know it's in RZ geom. 
`PeridynamicsMesh` doesn't know it's in RZ, does it? `Assembly` and `FEProblem`
have this info. It's more this problem that I am stuck with. 

If people think it's the better way, I can improve libmesh instead of shoe-horning it into MOOSE.

Reply:
We can access `FEProblem` and obtain the info, then pass it to the function, right? Or we can make this function as static in Assembly if the function accepts a Elem pointer and the coordinate type. I am just throwing rocks here ;-)